tests/integration/test_dbapi_integration.py

```diff
@@ -123,22 +124,114 @@ def test_string_query_param(trino_connection):
     assert rows[0][0] == "six'"
 
 
+def test_decimal_query_param(trino_connection):
+    cur = trino_connection.cursor()
+
+    cur.execute("SELECT ?", params=(Decimal('0.142857'),))
+    rows = cur.fetchall()
+
+    assert rows[0][0] == Decimal('0.142857')
+
+
 def test_datetime_query_param(trino_connection):
     cur = trino_connection.cursor()
 
-    cur.execute("SELECT ?", params=(datetime(2020, 1, 1, 0, 0, 0),))
+    params = datetime(2020, 1, 1, 16, 43, 22, 320000)
+
+    cur.execute("SELECT ?", params=(params,))
     rows = cur.fetchall()
 
-    assert rows[0][0] == "2020-01-01 00:00:00.000"
+    assert rows[0][0] == params
+    assert cur.description[0][1] == "timestamp"
+
+
+def test_datetime_with_time_zone_query_param(trino_connection):
+    cur = trino_connection.cursor()
+
+    params = datetime(2020, 1, 1, 16, 43, 22, 320000, tzinfo=pytz.timezone('CET'))
 
     cur.execute("SELECT ?",
-                params=(datetime(2020, 1, 1, 0, 0, 0, tzinfo=pytz.utc),))
+                params=(params,))
     rows = cur.fetchall()
 
-    assert rows[0][0] == "2020-01-01 00:00:00.000 UTC"
+    assert rows[0][0] == params
     assert cur.description[0][1] == "timestamp with time zone"
 
 
+def test_datetime_with_time_zone_numeric_offset(trino_connection):
+    cur = trino_connection.cursor()
+
+    cur.execute("SELECT TIMESTAMP '2001-08-22 03:04:05.321 -08:00'")
+    rows = cur.fetchall()
+
+    assert rows[0][0] == datetime.strptime("2001-08-22 03:04:05.321 -08:00", "%Y-%m-%d %H:%M:%S.%f %z")
+
+
+def test_special_datetimes_query_param(trino_connection):
+    cur = trino_connection.cursor()
+
+    for special_date in (
+            datetime.fromtimestamp(1603589478, pytz.timezone('Europe/Warsaw')),
+    ):
+        params = special_date
+
+        cur.execute("SELECT ?", params=(params,))
+        rows = cur.fetchall()
+
+        assert rows[0][0] == params
+
+
+def test_date_query_param(trino_connection):
+    cur = trino_connection.cursor()
+
+    params = datetime(2020, 1, 1, 0, 0, 0).date()
+
+    cur.execute("SELECT ?", params=(params,))
+    rows = cur.fetchall()
+
+    assert rows[0][0] == params
+
+
+def test_special_dates_query_param(trino_connection):
+    cur = trino_connection.cursor()
+
+    for params in (
+        # datetime(-1, 1, 1, 0, 0, 0).date(),
```

Comment:
If these are not supported, add a comment explaining why - is it because of some limitations in Python or other issues. If it's not relevant, remove these.

Reply:
We don't want to wrap these type of errors into a custom exception? Now the code will give a standard python exception.

Reply:
add inline comment beside each date with it's significance. See https://github.com/trinodb/trino/blob/7e99818f0194e6d792e3d67fc235511af279cf94/plugin/trino-sqlserver/src/test/java/io/trino/plugin/sqlserver/BaseSqlServerTypeMapping.java#L320 for inspiration.